hoodie-client/src/main/java/com/uber/hoodie/io/HoodieAppendHandle.java

```diff
@@ -151,11 +153,24 @@ private void init(Iterator<HoodieRecord<T>> recordItr) {
     }
 
     public void doAppend() {
+        List<Pair<String, Optional<IndexedRecord>>> allRecords =
+                records.stream().map(record -> Pair.of(record.getKey().getRecordKey(), getIndexedRecord(record)))
+                                .collect(Collectors.toList());
+
         List<IndexedRecord> recordItr =
-            records.stream().map(this::getIndexedRecord).filter(Optional::isPresent)
-                .map(Optional::get).collect(Collectors.toList());
+                allRecords.stream().map(Pair::getValue).filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .collect(Collectors.toList());
+
+        String [] keysToDelete =
```

Comment:
we are making 3 passes over this list - each of which allocates and deallocates list over again.. is there a way to use Collectors.groupBy or similar to just collect the two groups in pass?